[**TABLE OF CONTENTS**](toc.md)

[<<< Previous](14.md)   \|   [Next >>>](16.md)

Problem 15: Is vector exception-safe?
=====================================
Consider:
```c++
template <tyopename T> class vector {
	size_t n, cap;
	T *theVector;
public:
	vector(size_t n, const T&x): n{n}, cap{n}, 
		theVector{static_cast<T*>(operator new(n*sizeof(T)))} {
		for (size_t i = 0; i < n; ++i) {
			new(theVector + i) T (x);
			// copy constructor for T could throw, then what?
		}
	}
};
```
- Partially constructed vector - the destructor will not run.
- Broken invariant - don't have n valid objects

Note: if operator new throws: nothing has been allocated; no problem, strong guarantee

Fix:
```c++
template <typename T> vector<T>::vector (size_t n, const T&x):
	n{n}, cap{n}, theVector{static_cast<T*>(operator new(n*sizeof(T)))} {
		size_t progress = 0;
		try{
			for (size_t i = 0; i < n; ++i) {
				new(theVector + i) T(x);
				++progress;
			}
		} 
		catch (...) {	// catch anything
			for (size_t i = 0; i < progress; ++i) theVector[i].~T();
			operator delete(theVector);
			throw;		// rethrow
		}
	}
}
```

Abstract the filling part into its own function:
```c++
tempate <typename T> void uninitialized_fill(T* start, T* finish, const T&x) {
	T* p;
	try {
		for (p = start; p != finish, ++p) {
			new(static_Cast<void> (p)) T(x);
		}
	} 
	catch (...) {
		for (T*q = start; q!=p; ++q) q->~T();
		throw
	}
}

template <typename T> vector<T>::vector(size_t n, const T&x):
	n{n}, cap{n}, theVector{static_cast<T*>(operator new(n*sizeof(T)))} {
	try { 
		uninitialized_fill(theVector, theVector+n, x);
	} 
	catch(...) {
		operator delete(theVector);
		throw;
	}
}
```

Can clean this up by using RAII on the array.
```c++
template <typename T> struct vector_base {
	size_t n, cap;
	T *v;
	vector_base(size_t n): n{n}, cap{n == 0 ? 1 : n},
		v{static_cast<T*>(operator new(n*sizeof(T)))} {}
	~vector_base() {operator delete(v);}
}

template <typename T> class vector {
	vector_base<t> vb; // cleaned up implicitly when vector is destroyed.
public:
	vector(size_t n, const T&x): vb{n} {
		uninitialized_fill(vb.v, vb.v+vb.n, x);
	}
	~vector() { destroy_elements(); }
};
```

Copy constructor:
```c++
template <typename T> vector<T>::vector(const vector &other): vb{other.size()} {
	uninitialized_copy(other.begin(), other.end(), vb.v);
	// similar to uninitialized_fill, details: exercise
}
```

Exception safety of Vector
==========================
Assignment: copy and swap is exception safe, because swap is no-throw.


<hr>
[<<< Previous](14.md)   |   [Next >>>](16.md)